File: spikes/python/bench/parse_one.py
"""Parse a single file in an isolated process.

Hostile input must be observable (hang, crash, clean rejection) rather than
fatal to the whole benchmark run, so each case runs in its own process.
"""
import json
import sys


def parse_pdf(p):
    from pypdf import PdfReader
    r = PdfReader(p)
    return sum(len(pg.extract_text() or "") for pg in r.pages)


def parse_docx(p):
    import docx
    d = docx.Document(p)
    n = sum(len(par.text) for par in d.paragraphs)
    for t in d.tables:
        for row in t.rows:
            n += sum(len(c.text) for c in row.cells)
    return n


def parse_xlsx(p):
    from openpyxl import load_workbook
    wb = load_workbook(p)
    n = 0
    for ws in wb.worksheets:
        for _ in ws.iter_rows():
            n += 1
    return n


def parse_csv(p):
    import csv
    csv.field_size_limit(10_000_000)
    n = 0
    with open(p, newline="", encoding="utf-8", errors="replace") as fh:
        for _ in csv.reader(fh):
            n += 1
    return n


TABLE = {"pdf": parse_pdf, "docx": parse_docx, "xlsx": parse_xlsx, "csv": parse_csv}

if __name__ == "__main__":
    kind, file = sys.argv[1], sys.argv[2]
    try:
        units = TABLE[kind](file)
        sys.stdout.write(json.dumps({"outcome": "ok", "units": units}))
    except BaseException as e:  # noqa: BLE001 - classifying, not handling
        sys.stdout.write(json.dumps({
            "outcome": "clean_error",
            "error": type(e).__name__,
            "message": str(e)[:200],
        }))

File: spikes/typescript/bench/parse-one.mjs
// Parses a single file in an isolated process so hangs/crashes on hostile
// input are observable rather than fatal to the whole benchmark run.
import { readFile } from 'node:fs/promises';
import { STD_FONTS } from './lib.mjs';

const [, , kind, file] = process.argv;

async function parsePdf(p) {
  const pdfjs = await import('pdfjs-dist/legacy/build/pdf.mjs');
  const data = new Uint8Array(await readFile(p));
  // isEvalSupported:false disables the PDF JS engine (GHSA-hq66-cqwq-w95j class).
  const doc = await pdfjs.getDocument({ data, isEvalSupported: false, useSystemFonts: false,
    standardFontDataUrl: STD_FONTS }).promise;
  let n = 0;
  for (let i = 1; i <= doc.numPages; i++) {
    const tc = await (await doc.getPage(i)).getTextContent();
    n += tc.items.length;
  }
  return n;
}

async function parseDocx(p) {
  const mammoth = (await import('mammoth')).default;
  const r = await mammoth.extractRawText({ buffer: await readFile(p) });
  return r.value.length;
}

async function parseXlsx(p) {
  const ExcelJS = (await import('exceljs')).default;
  const wb = new ExcelJS.Workbook();
  await wb.xlsx.readFile(p);
  let n = 0;
  wb.eachSheet((ws) => ws.eachRow(() => n++));
  return n;
}

async function parseCsv(p) {
  const { parse } = await import('csv-parse');
  const { createReadStream } = await import('node:fs');
  let n = 0;
  await new Promise((res, rej) => {
    createReadStream(p).pipe(parse({ relax_column_count: true }))
      .on('data', () => n++).on('end', res).on('error', rej);
  });
  return n;
}

const table = { pdf: parsePdf, docx: parseDocx, xlsx: parseXlsx, csv: parseCsv };

try {
  const units = await table[kind](file);
  process.stdout.write(JSON.stringify({ outcome: 'ok', units }));
} catch (e) {
  process.stdout.write(JSON.stringify({
    outcome: 'clean_error',
    error: e && e.constructor ? e.constructor.name : 'Error',
    message: String(e && e.message || e).slice(0, 200),
  }));
}
